test: expect transaction to fail in http mode

Jolg42 opened the pull request.

janpio commented on driver-adapters/neon-http-node-basic/index.js:
```diff
@@ -2,7 +2,8 @@ const { Prisma, PrismaClient } = require('@prisma/client')
 
-const connectionString = process.env.DRIVER_ADAPTERS_NEON_HTTP_NODE_BASIC_DATABASE_URL
+const connectionString =
+  'postgres://prisma:pdYHi85xtOXa@ep-bold-meadow-77968096.us-east-1.aws.neon.tech/joel?connect_timeout=10'
```
ooopsie

Jolg42 replied:
Oopsie :D

Jolg42 replied:
Maybe we need to revoke this secret now... 🙈

janpio commented on driver-adapters/neon-http-node-basic/index.js:
```diff
@@ -22,134 +23,137 @@ exports.handler = async () => {
     const itxResult = await prisma.$transaction(getResult)
   } catch (e) {
-    if (e.message !== 'Transactions are not supported in HTTP mode') {
+    if (e.code !== 'P2036' && e.message !== 'PrismaClientKnownRequestError: Error in external connector (id 0)') {
```
Why no match on the error message from before? Huh?

Jolg42 replied:
"Transactions are not supported in HTTP mode" was what I expected without running the code

janpio replied:
Me as well, because that is what was output!?

Jolg42 replied:
See added test here https://github.com/prisma/ecosystem-tests/pull/4054/files#r1350495275

Jolg42 committed:
test: expect transaction to fail in http mode

Jolg42 committed:
Update schema.prisma

Jolg42 commented on driver-adapters/neon-http-node-basic/prisma/schema.prisma:
```diff
@@ -6,7 +6,7 @@ generator client {
   provider = "postgresql"
   // fake url to ensure the driver adapter is used
-  url      = "postgresql://foo"
+  url      = env("DRIVER_ADAPTERS_NEON_HTTP_NODE_BASIC_DATABASE_URL")
```
```suggestion
  url      = "postgresql://foo"
```

Jolg42 commented on driver-adapters/neon-http-node-basic/run.sh:
```diff
@@ -3,4 +3,4 @@
 
 pnpm install
-pnpm prisma generate
+pnpm prisma db push
```
```suggestion
pnpm prisma generate
```

Jolg42 committed:
Apply suggestions from code review

Jolg42 committed:
Merge branch 'dev' into joel/neon-http-transaction

Jolg42 committed:
wip

Jolg42 committed:
wip db push

Jolg42 committed:
add expectation & fix dependencies

Jolg42 committed:
require and extras ts-check

Jolg42 committed:
remove global jest

because it fails with `SyntaxError: Identifier 'jest' has already been declared`

see https://github.com/jestjs/jest/issues/9920

Jolg42 committed:
add orderBy on findMany

Jolg42 committed:
fix orderby (typo)

Jolg42 committed:
try catch around JSON.parse

Jolg42 commented on driver-adapters/neon-http-lambda-basic/index.js:
```diff
@@ -22,134 +24,162 @@ exports.handler = async () => {
-    statusCode: 200,
-    headers: { 'Content-Type': 'application/json' },
-    body: JSON.stringify(result),
+  // test the error message when the transaction fails
+  try {
+    await prisma.user.delete({
+      where: {
+        email: 'test-1@prisma.io',
+      },
+    })
+  } catch (e) {
+    if (e.code !== undefined && e.message !== 'Error: Transactions are not supported in HTTP mode') {
+      console.error('The expectation changed and needs to be updated in the test, see new error below:')
+      throw e
+    }
   }
+
+  // Transactions are expected to fail in HTTP mode
+  try {
+    const itxResult = await prisma.$transaction(getResult)
+  } catch (e) {
+    if (e.code !== 'P2036' && e.message !== 'PrismaClientKnownRequestError: Error in external connector (id 0)') {
+      console.error('The expectation changed and needs to be updated in the test, see new error below:')
+      throw e
+    }
+  }
```
iTx have a different error message?

Jolg42 replied:
TODO issue

Jolg42 replied:
https://github.com/prisma/team-orm/issues/431

millsp commented on driver-adapters/neon-lambda-basic/index.test.js:
```diff
@@ -14,7 +14,15 @@ test('prisma version and output', async () => {
-  const { regResult, itxResult } = JSON.parse(JSON.parse(new TextDecoder().decode(response.Payload)).body)
+  let regResult, itxResult
+  try {
+    const parsed = JSON.parse(JSON.parse(new TextDecoder().decode(response.Payload)).body)
+    regResult = parsed.regResult
+    itxResult = parsed.itxResult
+  } catch (e) {
+    console.log('new TextDecoder().decode(response.Payload)', new TextDecoder().decode(response.Payload))
+    throw e
```
nit: You could put the decoded result into a variable

millsp replied:
Also: does this change need to be ported to the other neon tests?

Jolg42 replied:
>nit: You could put the decoded result into a variable
I can't assign it to the non-local variable then

>Also: does this change need to be ported to the other neon tests?
yes, I'm opening a separate PR after merging this

Jolg42 replied:
see https://github.com/prisma/ecosystem-tests/pull/4092

millsp commented on driver-adapters/neon-http-lambda-basic/run.sh:
```diff
@@ -3,7 +3,6 @@
 set -eu
 
 pnpm install
-pnpm prisma generate
```
This was needed, I believe, because the dependencies are sometimes cached in the github job and that causes the client no to generate and then causing flakiness.

Jolg42 replied:
Looks like a typo, I didn't want to remove that one

millsp commented on driver-adapters/neon-http-lambda-basic/package.json:
```diff
@@ -7,6 +7,8 @@
   "devDependencies": {
     "@aws-sdk/client-lambda": "3.427.0",
+    "@jest/globals": "29.7.0",
+    "@jest/types": "^29.6.3",
```
I think renovate does not fix the carets atm, we can remove it manually

Jolg42 committed:
review fixes
